botcore/_bot.py

```diff
@@ -0,0 +1,265 @@
+            return None
+
+        self._remove_root_aliases(command)
+        return command
+
+    def clear(self) -> None:
+        """Not implemented! Re-instantiate the bot instead of attempting to re-use a closed one."""
+        raise NotImplementedError("Re-using a Bot object after closing it is not supported.")
+
+    async def on_guild_unavailable(self, guild: discord.Guild) -> None:
+        """Clear the internal guild available event when self.guild_id becomes unavailable."""
+        if guild.id != self.guild_id:
+            return
+
+        self._guild_available.clear()
+
+    async def on_guild_available(self, guild: discord.Guild) -> None:
+        """
+        Set the internal guild available event when self.guild_id becomes available.
+
+        If the cache appears to still be empty (no members, no channels, or no roles), the event
+        will not be set and `guild_available_but_cache_empty` event will be emitted.
+        """
+        if guild.id != self.guild_id:
+            return
+
+        if not guild.roles or not guild.members or not guild.channels:
+            msg = "Guild available event was dispatched but the cache appears to still be empty!"
+            self.log_to_dev_log(msg)
+            return
+
+        self._guild_available.set()
+
+    @abstractmethod
+    async def log_to_dev_log(self, message: str) -> None:
+        """Log the given message to #dev-log."""
+        ...
+
+    async def wait_until_guild_available(self) -> None:
+        """
+        Wait until the guild that matches the ``guild_id`` given at init is available (and the cache is ready).
+
+        The on_ready event is inadequate because it only waits 2 seconds for a GUILD_CREATE
+        gateway event before giving up and thus not populating the cache for unavailable guilds.
+        """
+        await self._guild_available.wait()
+
+    async def setup_hook(self) -> None:
+        """
+        An async init to startup generic services.
+
+        Connects to statsd, and calls
+        :func:`AsyncStatsClient.create_socket <botcore.async_stats.AsyncStatsClient.create_socket>`
+        and :func:`ping_services`.
+        """
+        loop = asyncio.get_running_loop()
+        self._connect_statsd(self.statsd_url, loop)
+        self.stats = AsyncStatsClient(loop, "127.0.0.1")
+        await self.stats.create_socket()
+
+        try:
+            await self.ping_services()
+        except Exception as e:
+            raise StartupError(e)
+
+    @abstractmethod
+    async def ping_services() -> None:
+        """Ping all required services on setup to ensure they are up before starting."""
+        ...
+
+    async def close(self) -> None:
+        """Close the Discord connection, and the aiohttp session, connector, statsd client, and resolver."""
+        # Done before super().close() to allow tasks finish before the HTTP session closes.
+        for ext in list(self.extensions):
```

Comment:
Why is this converted into a list?

Reply:
Never mind. It's because it unloads extensions, which would modify what was being iterated if a copy was not made.